### DISCUSSION TITLE
[1일차-지원] CSR vs SSR 차이

### DISCUSSION THREAD

**Discussion by smarfy99** [zzangCS/zzangInterview#3] [프론트 면접 질문]
CSR과 SSR의 차이를 말해보세요

**Comment by ez1n** [comment]
SSR : Server Side Rendering의 약자로 서버 쪽에서 렌더링 준비를 마친 상태로 클라이언트에 전달하는 방식

CSR : Client Side Rendering 의 약자로 초기 로드 시 모든 리소스들을 받아와 동적으로 DOM을 생성하여 렌더링하는 방식

차이점 : 
1. CSR의 경우 모든 HTML과 CSS, 스크립트를 한 번에 불러오고 SSR은 필요한 부분만 불러오기 때문에 SSR이 첫 페이지를 로딩하는 시간이 빠르다. 그러나 다른 페이지를 로딩하는 경우 SSR은 페이지별로 로딩을 다시 하기 때문에 CSR이 더 빠르다.
2. CSR은 초기에 HTML에 데이터가 없기 때문에 검색엔진최적화(SEO)에 취약할 수 있다.

**Reply by ez1n** [reply]
- CSR (Client Side Rendering)
    
> 서버에서 받은 단일 HTML 파일과 JS를 클라이언트(브라우저)에서 렌더링하는 방식
    
 서버에서 처리 없이 클라이언트로 보내주기 때문에 자바스크립트가 모두 다운로드 되고 실행이 끝나기 전까지 사용자는 화면을 볼 수 없다.
    
 **순서**

1. user가 website 요청을 보낸다.
2. CDN이 HTML 파일과 JS로 접근할 수 있는 링크를 클라이언트로 보낸다.
3. 클라이언트는 HTML과 JS를 다운로드 받는다. (화면을 볼 수 없음)
4. 다운로드 된 JS가 실행된다. (데이터를 위한 API 호출)
5. 서버가 API로부터의 요청에 응답한다.
6. API로부터 받아온 data를 plaeholder 자리에 넣어준다. (상호작용 가능)
    
- SSR (Server Side Rendering)
    
> 서버에서 렌더링 준비를 끝마친 상태로 클라이언트(브라우저)에 전달하는 방식
    
 서버에서 이미 렌더 가능한 상태로 클라이언트에 전달되기 때문에, JS가 다운로드 되는 동안 사용자는 화면을 볼수 있다.
    
**순서**

1. user가 website 요청을 보낸다.
2. server가 즉시 렌더링 가능한 html 파일을 만든다. (리소스 체크, 컴파일 후 완성된 HTML 컨텐츠로 만든다.)
3. HTML 파일이 클라이언트에 전달되면 즉시 렌더링 된다. (JS 읽히기 전)
4. 클라이언트가 자바스크립트를 다운받는다.
5. 브라우저가 JS 프레임워크를 실행한다.
6. 상호작용이 가능하다.

- 차이점
1. 페이지 로딩 시간
⇒ 첫 페이지 : CSR은 모든 페이지 소스를 한 번에 불러오고 SSR은 필요한 부분만 불러오기 때문에 평균적으로 SSR의 로딩 시간이 더 빠르다.
⇒ 나머지 페이지 : CSR은 처음에 모든 소스를 불러온 상태이고 SSR은 페이지가 이동할 때마다 로딩 과정을 재실행 하기 때문에 CSR의 로딩 시간이 더 빠르다. (= SSR의 로딩 시간이 느리다.)

2. SEO
⇒ 검색 엔진은 ‘크롤러’로 웹사이트를 읽어오는데 CSR은 자바스크립트 파일을 통해 동적으로 컨텐츠가 생성되기 때문에 처음에 빈 html파일로 존재하여 SEO에 취약하지만 SSR은 서버 사이드에서 컴파일 되어 클라이언트로 넘어오기 때문에 SEO에 유리하다.
⇒ 현재 구글의 검색엔진은 JS까지 실행시켜 판단함

3. 서버 자원 사용
⇒ SSR은 페이지가 바뀔 때마다 서버에 요청을 하기 때문에 서버 자원을 CSR에 비해 더 많이 사용한다.

**Comment by smarfy99** [comment]
SSR은 서버 쪽에서 기본적으로 필요한 html 코드를 미리 그려준 상태로 클라이언트에 전달하는 방식이다. 
CSR은 클라이언트 측에서 랜더링 과정을 통해 html을 채워가는 방식이다.
초기 로딩 시에 SSR은 서버 측에서 미리 그려진 html을 주기 때문에 로딩 속도가 빠르고, 검색엔진이 해당 콘텐츠를 크롤링하여 SEO에도 유리하다. 반면 초기 로딩 이후 다른 페이지로 넘어갔을 때, SSR은 리로딩을 해야함으로 CSR보다 속도가 느려질 수 있다.

**Comment by skyhanull** [comment]
csr은 클라이언트 사이드 랜더링의 약자로 클라이언트 단에서 랜더링을 하는 것이고 ssr은 서버사이드 랜더링의 약자로 서버단에서 랜더링을 하는 것을 말합니다 
이 둘의 차이점으로 초기로딩속도는 csr이 ssr보다 느린데 이유는 csr은 html,css,js를 초기로딩 때 다운을 받기에 느리지만 ssr은 필요한 부분만 데이터를 받아 랜더링하기에 csr보다 빠릅니다

나머지 로딩 시간은 반대로 csr은 이미 초반에 다 불러왓기 때문에 시간이 더 짧고 ssr은 요청할 때마다 필요한 데이터를 가져와야 하기에 느립니다.

seo(검색엔진 최적화)대응은 csr은 최초로 불러온 html이 js가 오기전까지 비어있어서 seo측면에서 좋지않고 ssr은 반대로 seo에 최적화 되어있습니다.
